File: lib/core/security/backup_exclusion.dart
import 'dart:io' show Directory, Platform;

import 'package:path_provider/path_provider.dart';

import '../../src/rust/api/os_security.dart' as rust_os;
import '../../utils/logger.dart';

/// Opt the app-support directory out of Apple's backup paths.
///
/// On iOS the application sandbox is indexed by iCloud Backup and by
/// encrypted iTunes/Finder backups by default. On macOS the same path
/// (`~/Library/Application Support/<bundle_id>`) is picked up by Time
/// Machine. Either of those turns the encrypted SQLite file, the KDF
/// parameters, the hardware-vault blob, and the password rate-limiter
/// journal into long-lived copies outside the device's trusted boot
/// chain — a restored backup on an attacker-controlled device gives
/// them offline brute-force time against the master password without
/// the per-device hardware binding the live install would have.
///
/// Both platforms honour the same API: setting
/// `URLResourceKey.isExcludedFromBackupKey` on a directory URL. On iOS
/// this maps to the `NSURLIsExcludedFromBackupKey` attribute (skipped
/// by iCloud + iTunes backup). On macOS the same resource-value call
/// writes the `com.apple.metadata:com_apple_backup_excludeItem`
/// extended attribute, which Time Machine honours.
///
/// Applied once on startup — idempotent, so re-running is cheap and
/// will refresh the flag if a system process stripped the xattr.
///
/// No-op on Android / Linux / Windows. Android's auto-backup exclusion
/// is handled at the manifest level via `data_extraction_rules.xml`;
/// Linux and Windows have no OS-level cloud-backup default the app
/// needs to opt out of.
///
/// Routes through `lfs_os_security::backup_exclusion::exclude_from_backup`
/// (objc2 → `NSURL.setResourceValue:forKey:NSURLIsExcludedFromBackupKey`).
/// The Rust function is a documented no-op on non-Apple targets, so the
/// Dart [_isApplePlatform] gate is a startup-path short-circuit only,
/// not a correctness guard.
class BackupExclusion {
  BackupExclusion({
    bool? isApplePlatform,
    Future<Directory> Function()? supportDir,
    void Function(String path)? excludeImpl,
  }) : _isApplePlatform =
           isApplePlatform ?? (Platform.isIOS || Platform.isMacOS),
       _supportDir = supportDir ?? getApplicationSupportDirectory,
       _excludeImpl = excludeImpl ?? _defaultExcludeImpl;

  final bool _isApplePlatform;
  final Future<Directory> Function() _supportDir;
  final void Function(String path) _excludeImpl;

  /// Flag the app-support directory so Apple backup paths skip it.
  /// Resolves to a no-op on non-Apple platforms.
  Future<void> applyOnStartup() async {
    if (!_isApplePlatform) return;
    try {
      final dir = await _supportDir();
      _excludeImpl(dir.path);
    } catch (e) {
      AppLogger.instance.log(
        'BackupExclusion.applyOnStartup failed: $e',
        name: 'BackupExclusion',
      );
    }
  }

  static void _defaultExcludeImpl(String path) {
    rust_os.osSecurityExcludeFromBackup(path: path);
  }
}

File: rust/crates/lfs_os_security/src/backup_exclusion.rs
//! Apple `URLResourceKey.isExcludedFromBackup` writer.
//!
//! Replaces the Swift `BackupExclusionPlugin` (`macos/Runner/`,
//! `ios/Runner/`) one-call `URL.setResourceValues` invocation.
//! On macOS this writes the
//! `com.apple.metadata:com_apple_backup_excludeItem` extended
//! attribute, which Time Machine honours; on iOS it sets the
//! private flag iCloud Backup / iTunes / Finder backup all
//! consult.
//!
//! Compile-time no-op on every non-Apple target (Linux, Windows,
//! Android) — those platforms either don't have a cloud-backup
//! default to opt out of, or the exclusion lives in a separate
//! manifest channel (Android `data_extraction_rules.xml`).
//!
//! Idempotent — re-running on a directory that already carries
//! the flag is a cheap no-op the OS handles internally. Failures
//! return a structured `Err(String)` so the Dart caller can log
//! without crashing the startup path.

#[cfg(any(target_os = "macos", target_os = "ios"))]
use objc2::msg_send;
#[cfg(any(target_os = "macos", target_os = "ios"))]
use objc2_foundation::{NSError, NSNumber, NSString, NSURL};

/// Set `NSURLIsExcludedFromBackupKey = true` on the directory at
/// `path`. Idempotent. Returns `Err` only when the OS surface
/// reports a real failure (path missing, permission denied,
/// filesystem doesn't support resource keys). Non-Apple targets
/// short-circuit to `Ok(())` so the caller doesn't have to gate
/// on `cfg!`.
#[cfg(any(target_os = "macos", target_os = "ios"))]
pub fn exclude_from_backup(path: &str) -> Result<(), String> {
    // SAFETY: every objc2 call here either returns an autoreleased
    // Retained<T> (memory-managed by the bindings) or sends a
    // documented Cocoa selector; no raw pointer arithmetic. The
    // unsafe blocks scope the `msg_send!` macros, which call
    // through to dynamic-dispatched selectors.
    unsafe {
        let path_ns = NSString::from_str(path);
        let url = NSURL::fileURLWithPath(&path_ns);
        let true_value = NSNumber::numberWithBool(true);
        let key = NSString::from_str("NSURLIsExcludedFromBackupKey");
        let mut error: *mut NSError = std::ptr::null_mut();
        let success: bool = msg_send![
            &*url,
            setResourceValue: &*true_value,
            forKey: &*key,
            error: &mut error,
        ];
        if success {
            return Ok(());
        }
        if !error.is_null() {
            let err_ref = &*error;
            let desc: *const NSString = msg_send![err_ref, localizedDescription];
            if !desc.is_null() {
                let s = (*desc).to_string();
                return Err(format!("setResourceValue failed: {s}"));
            }
        }
        Err("setResourceValue failed (unknown error)".to_string())
    }
}

/// Non-Apple targets: documented no-op so the Dart caller can
/// invoke the function unconditionally on startup without a
/// platform branch.
#[cfg(not(any(target_os = "macos", target_os = "ios")))]
pub fn exclude_from_backup(_path: &str) -> Result<(), String> {
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn no_op_on_non_apple_targets_does_not_error() {
        // The function compiles on every host — on Linux / Windows
        // it short-circuits to Ok. On macOS the call would attempt
        // a real setResourceValue against the path; we hand it a
        // path that almost certainly doesn't exist so the assertion
        // is "doesn't panic" rather than "succeeds".
        let result = exclude_from_backup("/tmp/lfs_backup_exclusion_no_such_path");
        if cfg!(not(any(target_os = "macos", target_os = "ios"))) {
            assert!(result.is_ok());
        }
        // Apple side: either Ok (file existed somehow) or a
        // structured error; both are acceptable here.
    }
}
